lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -27,6 +130,287 @@
+    count = 0
+    for relation in relations:
+        if name == relation.data[relation.app].get("network-name"):
+            count += 1
+
+    if count > 1:
+        logger.error(
+            f"There are multiple relations with the name {name}. Please change one or provide a custom network name."
+        )
+        return True
+    return False
+
+
+class RouterProvides(Object):
+    """This class is used to manage the routing table of the router provider.
+
+    It's capabilities are to:
+    * Manage the routing table in the charm itself, by adding and removing
+    new network and route requests by integrated units,
+    * Synchronize the databags of all requiring units with the router table of the
+    provider charm.
+
+    Attributes:
+        charm:
+            The Charm object that instantiates this class.
+        relationship_name:
+            The name used for the relationship implementing the ip-router interface
+            "ip-router" by default.
+    """
+
+    on = RouterProviderCharmEvents()
+
+    def __init__(self, charm: CharmBase, relationship_name: str = "ip-router"):
+        super().__init__(charm, relationship_name)
+        self.charm = charm
+        self.relationship_name = relationship_name
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._router_relation_changed
+        )
+        self.framework.observe(
+            charm.on[relationship_name].relation_departed, self._router_relation_departed
+        )
+
+    def _router_relation_changed(self, event: RelationChangedEvent):
+        """Resync the databags since there could have been a change in networks."""
+        if not self.charm.unit.is_leader():
+            return
+        self._sync_routing_tables()
+        self.on.routing_table_updated.emit()
+
+    def _router_relation_departed(self, event: RelationDepartedEvent):
+        """If an application has completely departed the relation, remove it
+        from the routing table.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        self._sync_routing_tables()
+
+    def get_routing_table(self) -> RoutingTable:
+        """Build the routing table from all of the related databags. Relations
+        that don't have missing or invalid network requests will be ignored.
+        """
+        logger.debug("Routing Table generation starting.")
+        router_relations = self.model.relations[self.relationship_name]
+        final_routing_table: RoutingTable = {}
+        for relation in router_relations:
+            new_network_name = relation.data[relation.app].get("network-name", None)
+            new_network_request: List[Network] = json.loads(
+                relation.data[relation.app].get("networks", "{}")
+            )
+
+            logger.debug(
+                f"Attempting to add ({new_network_request}) from app:({relation.app.name}) with relation-name:({new_network_name})"
```

Comment:
We avoid using fstrings for log messages.
Please use:
```suggestion
                "Attempting to add (%s) from app:(%s) with relation-name:(%s)", new_network_request, relation.app.name, new_network_name
```

Reply:
Done